File: n_fibo_ways.m
# Started since 3 Mar, 2018
# Author : anbarief@live.com

ex_1='1. Using Fibonacci function in a for-loop.';
disp(ex_1)
disp(fibo_1(16))

ex_2='2. Using Fibonacci formula fn=fn-1+fn-2 in a for-loop. ';
disp(ex_2)
disp(fibo_2(16))

ex_3='3. Similar as no.2, but using while-loop.';
disp(ex_3)
disp(fibo_3(16))

ex_4='4. Using recursive method.';
disp(ex_4)
disp(fibo_4(16))

ex_5='5. Using the sum function in a for-loop.';
disp(ex_5)
disp(fibo_5(16))

ex_6='6. Using Fibonacci function with vector input.';
disp(ex_6)
disp(fibo_6(16))

ex_7='7. Using Fibonacci alternative formula, the summation difference.';
disp(ex_7)
disp(fibo_7(16))

ex_8='8. Using a string to contain and generate Fibonacci numbers.';
disp(ex_8)
disp(fibo_8(16))

ex_9='9. Using the Fibonacci matrix form.';
disp(ex_9)
disp(fibo_9(16))




function f=fibo_1(n)
    
 for i=0:n-1
     f(i+1)=fibonacci(i);
 end
 
end
    
    
function f=fibo_2(n)
    
 f=[0,1];

 for i=3:n
    f(i)=f(i-1)+f(i-2);
 end

end


function f=fibo_3(n)
    
 f=[0,1];
 i=3;

 while (i<n+1)
    f(i)=f(i-1)+f(i-2);
    i=i+1;
 end

end


function f=fibo_4(n)
    
 f=[0,1];

 function y=recursive(m, a, b)
    
     if (m==3)
        y=a+b;
     else
        y=recursive(m-1, a+b, a);
     end
    
end

 for i=3:n
     f(i)=recursive(i, 1, 0);
 end

end


function f=fibo_5(n)
    
 f=[0,1];

 for i=3:n
    f(i)=sum(f(i-2:i-1));
 end

end


function f=fibo_6(n)
    
  f=fibonacci(0:n-1);
 
end


function f=fibo_7(n)
    
 f=[0,1];

 for i=3:n
     
     if (i>3)
        f(i)=sum(f)-sum(f(1:i-3));
     else
         f(i)=sum(f);
     end
     
 end

end


function f=fibo_8(n)
    
  f = '0-1-1';
  
  if n>3
   
      for i=4:n     
         logical = f=='-';
         strip = find(logical == 1);
         idx1 = strip(end);  idx2 = strip(end-1);
         a=str2num(f(idx1+1:end)); b=str2num(f(idx2+1:idx1-1));
         f(end+1)='-'; 
         new=num2str(a+b);
         f(end+2:end+2+length(new)-1)=new;
      end
     
  end 
 
end


function f= fibo_9(n)
   
    f=[0, 1];
  
    if n>3
        
        for i=3:n
            M = [1 1 ; 1 0]^(i-2);
            f(i) = M(1,1);
        end
        
    end
    
end

File: n_fibo_ways.cpp
// Author : anbarief@live.com
// Since 2 March 2018

#include <iostream>
#include <string>
#include <algorithm>
#include <iterator>
#include <vector>
#include <cmath>
#include <functional>

void fibo_1(int n){
	
	int f[n];
	
	f[0] = 0; f[1] = 1;

	std::cout << f[0] << "-" << f[1];
	
	for (int index = 2; index < n; index++){
		
		f[index] = f[index-1] + f[index-2];
		std::cout << "-" << f[index];
		
	}

}

void fibo_2(int n){
	
	int f[n]; int index=2;
	
	f[0] = 0; f[1] = 1;

	std::cout << f[0] << "-" << f[1];
	
	while (index < n) {
		
		f[index] = f[index-1] + f[index-2];
		std::cout << "-" << f[index];
		index = index + 1;
		
	}

}

int fibo_3(int n, int a = 0, int b = 1){
	
    std::cout << a << "-";
    
    if (n == 2){
    
    std::cout << b;
    
    return 0;
    	
    }
    
    return fibo_3(n-1, b, a+b);

}

void fibo_4(int n){
	
	std::vector<int> f(n, 0);
	f[1] = 1;
	
	std::cout << f[0] << "-" << f[1];
	
	for (int index = 2; index < n; index++){
		
		f[index] = f[index-1] + f[index-2];
		std::cout << "-" << f[index];
		
	}
	
}

void fibo_5(int n){
	
	std::vector<int> f(2, 0);
	f[1] = 1;
	
	std::cout << f[0] << "-" << f[1];
	
	for (int index = 2; index < n; index++){
		
		int val = f[index-1] + f[index-2];
		
		f.push_back(val);
		std::cout << "-" << val;
		
	}
	
}

void fibo_6(int n){
	
	std::vector<int> f(2, 0);
	f[1] = 1;
	
	std::cout << f[0] << "-" << f[1];
	
	for (int index = 2; index < n; index++){
		
		int val = f[index-1] + f[index-2];
		
		f.push_back(val);
		std::cout << "-" << f.back();
		
	}
	
}

void fibo_7(int n){
	
	float f[n];
	float sqr5 = std::sqrt(5);
	f[0] = 0; f[1] = 1;
	
	std::cout << f[0] << "-" << f[1];
	
	for (int index = 2; index < n; index++){
		
		f[index] = (1/sqr5)*(pow(0.5,index))*(pow(1+sqr5,index) - pow(1-sqr5,index));
		
		std::cout << "-" << f[index];
		
	}
	
	
}

void fibo_8(int n){
	
	float f[n];
	float sqr5 = std::sqrt(5);
	f[0] = 0; f[1] = 1;
	
	std::cout << f[0] << "-" << f[1];
	
	for (int index = 2; index < n; index++){
	
	if (index%2==0){
		f[index] = (1/sqr5)*(pow(0.5,index))*(pow(1+sqr5,index) - pow(1-sqr5,index));
	}
	else{
		f[index] = f[index-1] + f[index-2];
	};
		std::cout << "-" << f[index];
		
	}
	
}

std::vector<int> matrix_mul(int A[2][2], int B[2][2]){
	std::vector<int> res(4,  0);

	res[0] = A[0][0]*B[0][0] + A[0][1]*B[1][0];
    res[1] = A[0][0]*B[0][1] + A[0][1]*B[1][1];
    res[2] = A[1][0]*B[0][0] + A[1][1]*B[1][0];
    res[3] = A[1][0]*B[0][1] + A[1][1]*B[1][1];
    
	return res;
}

void fibo_9(int n){
	
	float f[n];
	int A[2][2], B[2][2];
	std::vector<int> M(4,0);
	
	f[0] = 0; f[1] = 1; f[2] = 1;
	
	A[0][0] = 1; A[0][1] = 1;
	A[1][0] = 1; A[1][1] = 0;
	
	B[0][0] = 1; B[0][1] = 1;
	B[1][0] = 1; B[1][1] = 0;
	
	
	std::cout << f[0] << "-" << f[1] << "-" << f[2];
	
	for (int index = 1; index < n-2; index++){
	// 2x2 Matrix multiplication 
    	M = matrix_mul(B, A);
    	B[0][0] = M[0]; B[0][1] = M[1];
    	B[1][0] = M[2]; B[1][1] = M[3];
    	
    	f[3] = M[0];
    	std::cout << "-" << f[3]; 
    }

}


int main(){
	
	int n = 20;
	
	std::string ex_1 = "fibo_1 : Using simple for-loop and fn = fn-1 + fn-2.";
	std::cout << '\n' << ex_1 << '\n';
	fibo_1(n);
	
	std::string ex_2 = "fibo_2 : Similar as fibo_1, but with while-loop.";
	std::cout << '\n' << ex_2 << '\n';
	fibo_2(n);
	
	std::string ex_3 = "fibo_3 : Using recursive method, with backward n.";
	std::cout << '\n' << ex_3 << '\n';
	fibo_3(n);
	
	std::string ex_4 = "fibo_4 : Similar as fibo_1, but using vector as container.";
	std::cout << '\n' << ex_4 << '\n';
	fibo_4(n);
	
	std::string ex_5 = "fibo_5 : Similar as fibo_4, using f.push_back method to add new Fibo number.";
	std::cout << '\n' << ex_5 << '\n';
	fibo_5(n);
	
	std::string ex_6 = "fibo_6 : Similar as fibo_5, but using f.back() to show the element.";
	std::cout << '\n' << ex_6 << '\n';
	fibo_6(n);
	
	std::string ex_7 = "fibo_7 : Using the analytical Fibo formula.";
	std::cout << '\n' << ex_7 << '\n';
	fibo_7(n);
	
	std::string ex_8 = "fibo_8 : Using combination of both Analytical formula, and the ususal fn = fn-1 + fn-2.";
	std::cout << '\n' << ex_8 << '\n';
	fibo_8(n);
	
	std::string ex_9 = "fibo_9 : Using Matrix identity of Fibonacci numbers, Fn = A^(n) F_init.";
	std::cout << '\n' << ex_9 << '\n';
	fibo_9(n);
		
	return 0;
}
